### PR TITLE
Avoid crashing when applying trivial contracts

### LINKED ISSUE #5941: Consistent order of parameters across the various instrument_* functions in code_contracts.*
CBMC version: 5.25.0
Operating system: macOS Mojave
Exact command line resulting in the issue: N/A.
What behaviour did you expect: N/A.
What happened instead:

Some `instrument_*` instrument_* functions from `code_contracts.h` use `function_id` while others don't. Make sure we have consistency among the parameters of these functions. Suggested [here](https://github.com/diffblue/cbmc/pull/5538/files#r593151285).

### LINKED ISSUE #5950: Remove all get_namespace from code_contracts
CBMC version: 5.25.0
Operating system: macOS Mojave
Exact command line resulting in the issue: N/A.
What behaviour did you expect: N/A.
What happened instead: N/A.

Refactor: remove all `get_namespace`from `code_contracts.*`. This could require refactoring input values across functions.

### PR DESCRIPTION
**Pull request by feliperodri**
Resolves https://github.com/diffblue/cbmc/issues/5950.
Resolves https://github.com/diffblue/cbmc/issues/5941.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/contracts/trivial_contract_enforce/test.desc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--enforce-all-contracts
+^EXIT=10$
+^SIGNAL=0$
+^\[main.assertion.\d+\] line \d+ assertion foo\(\&n\) != 15\: FAILURE$
+^VERIFICATION FAILED$
+--
+--
+Check whether CBMC doesn't crash when enforcing trivial contracts on functions.
```
Comment (martin-cs) [inline regression/contracts/trivial_contract_enforce/test.desc]: I am not really sure what "trivial" means in this context.  Checking whether a pointer is `NULL` is not always trivial.

File: regression/contracts/trivial_contract_enforce/main.c [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#include <assert.h>
+
+int foo(int *x) __CPROVER_requires(x != (void *)0)
```
Comment (martin-cs) [inline regression/contracts/trivial_contract_enforce/main.c]: Any reason for not using `NULL`?

File: src/goto-instrument/contracts/contracts.h [hunk 3]
```diff
@@ -177,43 +175,38 @@ class code_contractst
     goto_programt::instructionst::iterator &ins_it,
     goto_programt &program,
     exprt &assigns,
-    const irep_idt &function_id,
     std::set<irep_idt> &freely_assignable_symbols,
     assigns_clauset &assigns_clause);
 
-  /// Creates a local variable declaration for each expression in operands,
```
Comment (martin-cs) [inline src/goto-instrument/contracts/contracts.h:186]: This doesn't fit with the commit's message.  It really isn't ensuring a consistent ordering of parameters.

**Review by martin-cs — APPROVED** [review]
Yes but in future please can the commit messages and the code be a little closer.  For some of the refactorings, it is not clear why you are doing them.

File: regression/contracts/trivial_contract_enforce/test.desc (hunk 1 above)
Comment (feliperodri) [reply regression/contracts/trivial_contract_enforce/test.desc]: I updated the test description. The "trivial" refers to contracts with only a requires clause, which would be basically adding an assumption. It's not related to nullness check.

File: regression/contracts/trivial_contract_enforce/main.c (hunk 2 above)
Comment (feliperodri) [reply regression/contracts/trivial_contract_enforce/main.c]: No. I added a NULL pointer.

File: src/goto-instrument/contracts/contracts.h (hunk 3 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/contracts.h:186]: Good point, Martin! I broke down the commit into three. I'll keep this in mind for future PRs.

File: regression/contracts/trivial_contract_enforce/test.desc [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+main.c
+--enforce-all-contracts
+^EXIT=10$
+^SIGNAL=0$
+^\[main.assertion.\d+\] line \d+ assertion foo\(\&n\) != 15\: FAILURE$
+^VERIFICATION FAILED$
+--
+--
+Check whether CBMC doesn't crash when enforcing incomplete contracts, i.e.,
+without requires using ensures and vice-versa.
```
Comment (SaswatPadhi) [inline regression/contracts/trivial_contract_enforce/test.desc]: I would reword this description a bit:
- In the test, we are not checking the other direction: "ensures without requires" but we say "vice-versa"
- The crash was not strictly due to "incomplete" contracts. It could be triggered by writing `__CPROVER_ensures(true)` as opposed to removing it entirely.

I would say exactly what the PR title says and define what "trivial" means in an enforce context, which is that the postcondition is true (default when missing) and therefore there is nothing to check/assert.

File: regression/contracts/trivial_contract_replace/test.desc [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+main.c
+--enforce-all-contracts
+^EXIT=10$
+^SIGNAL=0$
+^\[main.assertion.\d+\] line \d+ assertion foo\(\&n\) != 15\: FAILURE$
+^VERIFICATION FAILED$
+--
+--
+Check whether CBMC doesn't crash when replacing incomplete contracts, i.e.,
+without requires using ensures and vice-versa.
```
Comment (SaswatPadhi) [inline regression/contracts/trivial_contract_replace/test.desc]: Similar comment as above. In the test, we are not checking the other direction: "requires without ensures" but we say "vice-versa", so do you want to add another function for checking this direction as well?

**Review by SaswatPadhi — APPROVED** [review]
LGTM, with couple of minor comments:

File: regression/contracts/trivial_contract_enforce/test.desc (hunk 4 above)
Comment (feliperodri) [reply regression/contracts/trivial_contract_enforce/test.desc]: Implemented as suggested.

File: regression/contracts/trivial_contract_replace/test.desc (hunk 5 above)
Comment (feliperodri) [reply regression/contracts/trivial_contract_replace/test.desc]: Implemented as suggested.

File: doc/cprover-manual/contracts-assigns.md [hunk 6]
```diff
@@ -112,7 +114,7 @@ Clauses](contracts-requires-and-ensures.md) - Replacement section, and it will a
 non-deterministic assignments for each object listed in the `__CPROVER_assigns`
 clause. Since these objects might be modified by the function, CBMC uses
 non-deterministic assignments to havoc them and restrict their values only by
-assuming the postconditions.
+assuming the postconditions (i.e., requires clauses).
```
Comment (SaswatPadhi) [inline doc/cprover-manual/contracts-assigns.md:117]: Thanks for fixing the docs! So many tiny errors (including errors in code examples) went unnoticed the last time!

Just one small issues: I think you meant "ensures" (not "requires") clauses in this sentence.
